### ISSUE TITLE
Windows Agent Hides or Capitalizes Variables

### ISSUE THREAD

**Issue by dansiegel** [Microsoft/azure-pipelines-agent#1981]
## Having generic issue with Azure-Pipelines/VSTS/TFS?
See the original Developer Community Issue [here](https://developercommunity.visualstudio.com/content/problem/305743/build-variables-not-added-or-added-in-all-caps.html).

## Agent Version and Platform
Hosted VS2017 

## VSTS Type and Version
dev.azure.com

## What's not working?
Add the following variables to a pipeline... *note* this could be done either with unprotected variables or explicitly on a build step via YAML...

- Secret_Foo
- Manifest_Bar
- FooBar

Now run a Pipeline on a Hosted VS2017 agent with a powershell script to dump environment variables:

```ps1
gci env:* | sort-object name
```

Your output will not include any variation of `Secret_Foo`, and any other variables you have added to the environment will have been capitalized meaning you now have `MANIFEST_BAR` and `FOOBAR` instead of what you actually entered. This behavior is completely different on the Hosted Mac agent, where each variable is available and appears in the environment as entered into the pipeline.

| Variable | Hosted VS2017 | Hosted macOS |
|:--------:|:------------:|:--------:|
| Secret_Foo | n/a | Secret_Foo |
| Manifest_Bar | MANIFEST_BAR | Manifest_Bar |
| FooBar | FOOBAR | FooBar |

The above would be true whether working with an unprotected variable or added in YAML like the following:

```yml
- task SomeTask
  env:
    Secret_Foo: $(ProtectedVariable)
    Secret_Bar: $(UnprotectedVariable)
```

Also worth noting that this behavior is consistent across direct Azure DevOps Hosted pipelines and App Center Builds.

**Comment by TingluoHuang** [comment]
@dansiegel we choose to replace `.` and `toUpper` so you can reference the variable in the same format across both Windows, Linux and macOS. https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=vsts&tabs=yaml%2Cbatch#working-with-variables

`Secret_` is a reserved variable prefix, since task-lib will consider that is a secret variable, there are also `Input_`, `Endpoint_` etc.

the Windows behavior is what i expected, i would say that is by design.
I am not sure about the macOS behavior, can you provide more log about that, like how you set the variable and how did you print them out?

**Comment by bryanmacfarlane** [comment]
It's not about easier, it's about consistency across all platforms so folks can write tasks, release to the market place for all platforms.

I just ran a build on mac os with a bash script running `env` and it's consistent.

```bash
2018-11-21T12:30:55.6268810Z MANIFEST_BAR=some value
...
2018-11-21T12:30:55.6269810Z FOOBAR=somevalue
```

So I thought maybe it was a bug in running powershell core on a mac, so I tried an inline powershell step running

```bash
2018-11-21T12:42:40.1328170Z FOOBAR                         somevalue
...
2018-11-21T12:42:40.1376670Z MANIFEST_BAR                   some value
```

So I can't repro and it looks to be consistent as designed.

**Comment by dansiegel** [comment]
@TingluoHuang @bryanmacfarlane sorry guys, missed the response. 

I'm gonna have to reiterate @willmorgan here... 

>  I appreciate your answer on this, but "by design" doesn't make this OK.

Also:

>  so you can reference the variable in the same format across both Windows, Linux and macOS
 
That is a COMPLETELY FALSE statement. I wouldn't have opened the issue in the first place if the behavior were the same across environments... in fact the ONLY one that handles Variables wrong is Windows. If you designed it this way, your design is bad and needs to be fixed.

**Comment by hbuckle** [comment]
See also #1645 - this may be by design but in this case I agree that the design is hindering rather than helping. A lot of tools are cross platform now so preserving case sensitivity is important on all platforms.

**Comment by eXigentCoder** [comment]
This auto-uppercase is causing a big headache for us, as described below.

To help keep our applications config manageable, rather than just referring to `process.env.<variable-name>` we have a base config object which may or may not have some typing info based on a TS definition. This config object is built up using a combination of files checked in to the repo, arguments passed in to the running process and environment variables, especially for secrets.

An example of a config file that we use is as follows:
![config](https://user-images.githubusercontent.com/8137674/60011504-32903980-967a-11e9-951f-8675f95dd762.png)

[nconf](https://www.npmjs.com/package/nconf) then uses the convention of the double underscore `__` to take env and argv variables and set properties on the object for different environments. So some example environment variables related to the above file would be:
![env](https://user-images.githubusercontent.com/8137674/60011589-623f4180-967a-11e9-99bf-8e36ca4bbc73.png)

As you can probably figure out, now that all the env variables are automatically uppercased the intended properties no longer get set/replaced.
It would be really great to have the option to turn off this functionality for a pipeline or on a per-variable basis

**Comment by arbreezy** [comment]
It would be great if we can have a system variable to opt out of this upper case function in pipelines.

**Comment by dansiegel** [comment]
@arbreezy you're right it would be great if they would at least give us that compromise... but unfortunately @TingluoHuang @bryanmacfarlane both seem to think that the fact this issue was created specifically due to the inconsistency with other agents and the fact that they're breaking people doesn't matter... this issue really should be reopened and addressed. 

To be pretty blunt here, this team needs to start considering the needs of developers rather than making decisions on their own that have a track record of breaking other teams at Microsoft, and large community OSS projects in addition to lots of customers.

**Comment by willmorgan** [comment]
> It would be great if we can have a system variable to opt out of this upper case function in pipelines.

Yes! But only if it's called `OpT_out_Plz`.

**Comment by frarteaga** [comment]
@dansiegel I reach out this discussion because I have installed your Mobile.BuildTools package and is not working the secrets feature on my azure devops pipeline, now I know way. But I think the choose of `Secret_` prefix was a bad decision as exposed by @TingluoHuang:

> Secret_ is a reserved variable prefix, since task-lib will consider that is a secret variable, there are also Input_, Endpoint_ etc.

@TingluoHuang I also think that the reserved variable prefixes should be informed in the documentation. Now I wondering why is reserved in windows agents but not in linux and macos agents?
